### ISSUE TITLE
The default settings for SSH access need hardening

### ISSUE THREAD

**Issue by KorakSilvercloud** [freedombox/Plinth#706]
Currently the default settings create fairly big risk. SSH is exposed to the Internet on the default port 22. Root login is allowed, and there is nothing limiting log in attempts. I had mistakenly assumed that the firewall blocked remote connections and only allowed SSH from local network. Upon finding out it was exposed to the net I have checked my logs and found I am under a constant bruit force attack from a number of different attackers. 

This is fairly serious issue as many users do not use secure passwords that would stand up to a prolonged attack.

**Comment by jvalleroy** [comment]
Just to check, by root login, do you mean login by root user? Or login by user with sudo privileges (who can then become root)?

**Comment by KorakSilvercloud** [comment]
By root login I mean directly ssh in as root. Really if this project is intended for non-technical users then SSH should not be exposed to the internet in any form. I was lucky that I had selected a strong password.

**Comment by jvalleroy** [comment]
@KorakSilvercloud Are you using a pre-built image, or did you set up FreedomBox on an already installed Debian system?

**Comment by KorakSilvercloud** [comment]
I did a clean install of Debian and then followed directions to setup FreedomBox.

**Comment by jvalleroy** [comment]
@KorakSilvercloud Ok, it looks like this is actually the default configuration for openssh-server package.

https://sources.debian.net/src/openssh/1:7.4p1-6/debian/openssh-server.templates/
https://sources.debian.net/src/openssh/1:7.4p1-6/debian/openssh-server.postinst/

There's a debconf configuration named "permit-root-login", with description "Disable SSH password authentication for root?". It's default value is false, which apparently means "root password login is permitted". We should reconfigure the value to be true, which means "root password login is prohibited".

**Comment by KorakSilvercloud** [comment]
Yes, the default openssh-server configuration is I think meant more for internal servers. It is not what you want to expose to the internet.

Disabling root login I would consider a must as root is by far the most attacked login. I would also say having something like fail2ban would be necessary if you plan on keeping this exposed on the WAN side. Really I would say by default SSH should be LAN only.

Just to put this in perspective. I ran a honeypot on my home connection where I had been running a freedombox. In less than a week I had seen over 11,000 attempts against the box and the attack rate was accelerating.

**Comment by SunilMohanAdapa** [comment]
On Tuesday 24 January 2017 10:30 PM, KorakSilvercloud wrote:
> Yes, the default openssh-server configuration is I think meant more for
> internal servers. It is not what you want to expose to the internet.
> 
> Disabling root login I would consider a must as root is by far the most
> attacked login. I would also say having something like fail2ban would be
> necessary if you plan on keeping this exposed on the WAN side. Really I
> would say by default SSH should be LAN only.
> 
> Just to put this in perspective. I ran a honeypot on my home connection
> where I had been running a freedombox. In less than a week I had seen
> over 11,000 attempts against the box and the attack rate was accelerating.

We should have fail2ban by default.  The default configuration nicely
works for SSH without any modifications.  All we need to do is add this
as a dependency for our security module.  In future, we can also ship
fail2ban configuration for Plinth so that Plinth logins are also
controlled by fail2ban.

Note that as it stands, although root login is permitted by
openssh-server, it is not possible to login to FreedomBox as root
because password login is disabled.  We should change the openssh-server
configuration to disable root login just to be safe.

-- 
Sunil

**Comment by jvalleroy** [comment]
On 01/25/2017 02:22 AM, Sunil Mohan Adapa wrote:
> Note that as it stands, although root login is permitted by
> openssh-server, it is not possible to login to FreedomBox as root
> because password login is disabled. We should change the openssh-server
> configuration to disable root login just to be safe.

I think you are referring to the root password being disabled in our
images. But for someone following
https://wiki.debian.org/FreedomBox/Hardware/Debian, they may or may not
have set a root password during Debian install.

**Comment by KorakSilvercloud** [comment]
I guess I still just question, is there a solid and valid reason why you would keep SSH enabled on the WAN side?

The impression I kept getting from people is that this project is supposed to be building a system for non-technical people. If that is the case then your users are not exactly the type to ever use SSH, they certainly are not the type to be doing remote SSH connections. Exposing this to the internet just creates a larger attack surface. When building a system to help people increase online privacy it seems like security would be one of the top issues.

**Comment by aaronferrucci** [comment]
I also see many attempts to login via ssh to my freedombox. pam_abl is on by default, and blocks some of the attacks, but I think it's only a matter of time before the hackers' scripts start including 'fbx' in their username lists. Then anyone who chose a weak password will be vulnerable. Setting stronger sshd defaults would make sense to me - also how about the option to use a non-default port for sshd? I suspect that would help a lot.

**Comment by Raphi111** [comment]
Any update on this ? 
Although it's a completely new box, I also keep being hammered on the SSH port (one attempt every minute or so). 
I agree that as Freedombox also targets non-tech users, it should be far more secure on this port by default. I tend to use strong passwords by default, but imagine someone in a rush setting up very simple passwords like "admin-admin" or "user-password" just to test Freedombox. Then he decide to keep his box running endlessly because he likes the services it offers. He would be pwned within hours, if not minutes.
Or even worse, someone who installed Freedombox on top of Debian might have setup a poor root password, as someone mentioned.
Maybe SSH should be disabled on WAN right from the start, or Plinth should recommend to disable it once the user box is set up.
Only allowing SSH keys logins (disabling password logins) would also be more secure, maybe Plinth could offer a way to import public keys.
Also, even if it could prevent bots, I don't think changing the port alone would be an acceptable solution, as it's just "hiding" the box a little bit more, but any serious attacker would find it immediately.

Please don't ignore this issue, as it makes many default Freedomboxes an easy target.
